src/app/components/elements/inputs/PasswordInput.tsx

```diff
@@ -0,0 +1,108 @@
+import React, { useState } from "react";
+import {
+  FormFeedback,
+  FormGroup,
+  Input,
+  Label,
+  UncontrolledTooltip,
+} from "reactstrap";
+import { PasswordFeedback, ValidationUser } from "../../../../IsaacAppTypes";
+import { Immutable } from "immer";
+import { passwordDebounce, validatePassword } from "../../../services";
+
+interface PasswordInputProps {
+  fieldType: "password" | "confirmPassword";
+  userToUpdate: Immutable<ValidationUser>;
+  setUserToUpdate: (user: Immutable<ValidationUser>) => void;
+  setUnverifiedPassword: (password: string) => void;
+  submissionAttempted: boolean;
+  unverifiedPassword: string | undefined;
+  defaultPassword?: string;
+}
+
+export const PasswordInput = ({
+  fieldType,
+  userToUpdate,
+  setUserToUpdate,
+  unverifiedPassword,
+  setUnverifiedPassword,
+  submissionAttempted,
+  defaultPassword,
+}: PasswordInputProps) => {
+  const [passwordFeedback, setPasswordFeedback] =
+    useState<PasswordFeedback | null>(null);
+
+  const passwordIsValid =
+    userToUpdate.password == unverifiedPassword &&
+    validatePassword(userToUpdate.password || "");
+
+  const newPassword = (
+    <FormGroup>
+      <Label htmlFor="password-input">Password</Label>
+      <span id={`password-help-tooltip`} className="icon-help ml-1" />
+      <UncontrolledTooltip target={`password-help-tooltip`} placement="bottom">
+        {
+          "Passwords must be at least 12 characters, containing at least one number, one lowercase letter, one uppercase letter, and one special character."
+        }
+      </UncontrolledTooltip>
+      <Input
+        id="password-input"
+        type="password"
+        name="password"
+        autoComplete="new-password"
+        required
+        defaultValue={defaultPassword}
+        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
```

Comment:
doesn't onChange on a text input get called every time a character is added?

Reply:
Yes, hence the many re-renders going on in the background.  But I don't see a great way of avoiding it currently; we do want the password field to update visually for the user on every key press, and update the state to check against PW requirements.  I could put a short debounce to delay the action and reduce the frequency, but that may be perceptible to fast typers?